### ISSUE TITLE
train_draft.cpp — O(H³) matrix inversion on hidden-state-sized matrices is numerically unstable and prohibitively slow

### ISSUE THREAD

**Issue by bong-water-water-bong** [1bit-MONSTER/1bit-MONSTER#608]
## Summary

`tools/train_draft.cpp` performs Gauss-Jordan elimination to invert an H×H matrix where H is the model's hidden dimension (typically 1024-4096). This is O(H³) in both time and accumulated floating-point error, and for H=4096 means ~68 billion operations with catastrophic numerical precision loss.

## Location

`tools/train_draft.cpp`, lines 101-144:

```cpp
// Gauss-Jordan elimination
for(int col=0;col<H;col++){
    // Find pivot
    int pivot=col;
    double max_val=fabs(XtX[col*H+col]);
    for(int row=col+1;row<H;row++){
        double v=fabs(XtX[row*H+col]);
        if(v>max_val){max_val=v;pivot=row;}
    }
    if(max_val<1e-10){printf("  Singular matrix at col %d\n",col);continue;}
    ...
}
```

### 1. Numerical Instability
- Gauss-Jordan is one of the least stable matrix inversion methods
- For hidden dimension H=2048, the matrix is 2048×2048 = 4 million entries
- After 2000+ elimination steps, accumulated rounding errors dominate the result
- The `continue` on singular pivot silently produces garbage for that column
- Using `double` helps but doesn't fix the fundamental stability issue

### 2. Computational Cost
- O(H³) with H=2048 → ~8.5 billion FLOPS for a single inversion
- For H=4096 → ~68 billion FLOPS
- This runs single-threaded on CPU — could take minutes to hours

### 3. Memory
- Two H×H matrices of doubles: 2048² × 8 bytes × 2 = 64 MB (okay)
- For 4096: 4096² × 8 × 2 = 256 MB (less okay)

### 4. Incorrect Approach
- Learning a full H×H weight matrix from n_samples=1000 is severely underdetermined (H² >> n_samples for H > ~32)
- A linear transformation with 4M+ parameters trained on 1000 samples will massively overfit
- The method assumes the mapping from draft→full hidden states is linear, which it isn't

## Impact

- Training either crashes on singular matrices or produces a garbage adapter
- The validation metric (top-1 token match) showing improvement may be overfitting noise
- The entire approach is unlikely to produce a useful speculative decoding adapter

## Fix

Replace with a numerically stable alternative:
1. Use QR decomposition or SVD via LAPACK/BLAS instead of Gauss-Jordan
2. Use a smaller adapter (low-rank approximation, e.g., W = AB with rank r << H)
3. Or use PyTorch's `torch.linalg.lstsq` instead of manual C++ inversion

**Comment by bong-water-water-bong** [comment]
Fixed by #618 (merged) — replaced Gauss-Jordan with Cholesky decomposition.
